protobufs/src/tests.cpp

```diff
@@ -10,37 +10,115 @@
+#include "cdh.hpp"
+#include <map>
+#include <stdint.h>
 
-inline bool itDoesThis(){ return true; }
-inline bool itDoesThat(){ return true; }
-
-SCENARIO( "Do that thing with the thing", "[Tags]" ) {
-    GIVEN( "This stuff exists" ) {
-        // make stuff exist
-        WHEN( "I do this" ) {
-            // do this
-            THEN( "it should do this")
-            {
-                REQUIRE( itDoesThis() );
-                AND_THEN( "do that")
-                    REQUIRE( itDoesThat() );
-            }
-        }
+SCENARIO("Check that subsystems are fully defined", "[bdd][cdh][subsystems]") {
+  GIVEN("The subsystem enum") {
+    using namespace cdh::subsystems;
+    WHEN("iterating over the enum") {
+      THEN("the symmetry in numbering should be correct") {
+        int subsys_id = CDH;
+        do {
+          INFO("checking subsys_id = " << subsys_id << " out of " << HMI);
+          REQUIRE(subsystem_from_ID(subsys_id) == subsys_id);
+          AND_THEN(
+              "it should also have a std::string name and can be reversed");
+          REQUIRE(subsystem_from_name(subsystem_name(
+                      subsystem_from_ID(subsys_id))) == subsys_id);
+          subsys_id++;
+        } while (subsys_id < HMI);
+      }
     }
+  }
 }
 
-SCENARIO( "Test the telemetry", "[bdd][cdh][telemetry]" ) {
-    cdh::telemetry::TelemetryData TMTest;
-    TMTest.set_id(0x01);
-    TMTest.set_bool_value(true);
-    GIVEN( "A new TM point" ) {
-        // make stuff exist
-        WHEN( "I set its ID to 0x01 and its bool value to `true`" ) {
-            // do this
-            THEN( "its bool value should be set to true")
-            {
-                  REQUIRE( TMTest.bool_value() );
-            }
+SCENARIO("Mock telemetry", "[bdd][cdh][telemetry]") {
+  GIVEN("Some mock telemetry collection,") {
+    WHEN("Collecting all telemetry,") {
+      cdh::telemetry::TMFrame frame = cdh::subsystems::collect_all_telemetry();
+      THEN("the first seq_no should be 0,") {
+        REQUIRE(frame.sequence_no() == 0);
+      }
+      THEN("the second seq_no should be 1,") {
+        REQUIRE(frame.sequence_no() == 1);
+      }
+      THEN("the first management telemetry point should be a valid CDH frame") {
+        cdh::telemetry::Telemetry cdh_mgmt = frame.tm(0);
+        REQUIRE(cdh_mgmt.sys() == cdh::subsystems::CDH);
+        AND_THEN("its ID should 0x01");
+        REQUIRE(cdh_mgmt.id() == 0x01);
+        AND_THEN("the case of the data should be int");
+        REQUIRE(cdh_mgmt.data_case() ==
+                cdh::telemetry::Telemetry::DataCase::kIntValue);
+        AND_THEN("the max sequence number should be set to the limit of uint");
+        // NOTE: We convert this manually to a unit because the TM does not
+        // support unit, only int, but we know what this should be.
+        REQUIRE((uint)cdh_mgmt.int_value() == UINT_MAX);
+      }
+      THEN("each subsystem should have the same mock telemetry points") {
```

Comment:
Although seemingly complicated initially, I find this test pretty cool. My opinion may change later.